@qmuntal it looks like the `runtime` tests failed and the JUnit conversion didn't work... might want to take a look at `raw-test-output-linux_ubuntu_default_amd64_test_nosystemcrypto_False-retry-1`?

The test result upload failed with:

```
##[warning]Failed to read /mnt/vss/_work/1/s/eng/artifacts/TestResults.xml. Error :
Unexpected end of file has occurred. The following elements are not closed: testsuites. Line 102950, position 13..
```

Raw JSON shows the test failure like this:

```
{"Time":"2025-07-24T16:20:13.070634892Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"    runtime-gdb_test.go:524: GDB command timed out after 7m9.697211128s: /usr/bin/gdb -nx -batch -iex add-auto-load-safe-path /__w/1/s/go/src/runtime -ex set startup-with-shell off -ex break main.eee -ex run -ex backtrace -ex continue /tmp/TestGdbBacktrace3352410727/001/a.exe\n"}
{"Time":"2025-07-24T16:20:13.074532267Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"    runtime-gdb_test.go:529: gdb output:\n"}
{"Time":"2025-07-24T16:20:13.074548367Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        Loading Go Runtime support.\n"}
{"Time":"2025-07-24T16:20:13.074553667Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        Breakpoint 1 at 0x46fa20: file /tmp/TestGdbBacktrace3352410727/001/main.go, line 17.\n"}
{"Time":"2025-07-24T16:20:13.074559267Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        warning: Error disabling address space randomization: Operation not permitted\n"}
{"Time":"2025-07-24T16:20:13.074564367Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        [New LWP 60774]\n"}
{"Time":"2025-07-24T16:20:13.074568867Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        [New LWP 60775]\n"}
{"Time":"2025-07-24T16:20:13.074573168Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        [New LWP 60776]\n"}
{"Time":"2025-07-24T16:20:13.074577668Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        [New LWP 60777]\n"}
{"Time":"2025-07-24T16:20:13.074582868Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        \n"}
{"Time":"2025-07-24T16:20:13.074588168Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        Thread 1 \"a.exe\" hit Breakpoint 1, main.eee (~r0=\u003coptimized out\u003e) at /tmp/TestGdbBacktrace3352410727/001/main.go:17\n"}
{"Time":"2025-07-24T16:20:13.074594168Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        17\tfunc eee() bool { return true }\n"}
{"Time":"2025-07-24T16:20:13.074604068Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        #0  main.eee (~r0=\u003coptimized out\u003e) at /tmp/TestGdbBacktrace3352410727/001/main.go:17\n"}
{"Time":"2025-07-24T16:20:13.074609468Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        #1  0x000000000046fa16 in main.ddd (~r0=\u003coptimized out\u003e) at /tmp/TestGdbBacktrace3352410727/001/main.go:14\n"}
{"Time":"2025-07-24T16:20:13.074614468Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        #2  0x000000000046f9ef in main.ccc (~r0=\u003coptimized out\u003e) at /tmp/TestGdbBacktrace3352410727/001/main.go:11\n"}
{"Time":"2025-07-24T16:20:13.074619368Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        #3  0x000000000046f9cf in main.bbb (~r0=\u003coptimized out\u003e) at /tmp/TestGdbBacktrace3352410727/001/main.go:8\n"}
{"Time":"2025-07-24T16:20:13.074624369Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        #4  0x000000000046f9af in main.aaa (~r0=\u003coptimized out\u003e) at /tmp/TestGdbBacktrace3352410727/001/main.go:5\n"}
{"Time":"2025-07-24T16:20:13.074629269Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        #5  0x000000000046fa4f in main.main () at /tmp/TestGdbBacktrace3352410727/001/main.go:22\n"}
{"Time":"2025-07-24T16:20:13.074633869Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        [LWP 60775 exited]\n"}
{"Time":"2025-07-24T16:20:13.074638169Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        [LWP 60774 exited]\n"}
{"Time":"2025-07-24T16:20:13.074642569Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        [LWP 60771 exited]\n"}
{"Time":"2025-07-24T16:20:13.074646669Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"        [LWP 60777 exited]\n"}
{"Time":"2025-07-24T16:20:13.07468817Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"    runtime-gdb_test.go:549: gdb exited with error: signal: killed\n"}
{"Time":"2025-07-24T16:20:13.075476385Z","Action":"output","Package":"runtime","Test":"TestGdbBacktrace","Output":"--- FAIL: TestGdbBacktrace (430.36s)\n"}
{"Time":"2025-07-24T16:20:13.075497385Z","Action":"fail","Package":"runtime","Test":"TestGdbBacktrace","Elapsed":430.36}
```

Not sure if something in this change is causing the conversion to fail upon test failure. 😕

--- a/eng/_util/buildutil/testjson.go
+++ b/eng/_util/buildutil/testjson.go
@@ -26,7 +26,7 @@ type TestJSONFlags struct {
 func BindTestJSONFlags() *TestJSONFlags {
 	var f TestJSONFlags
 	flag.StringVar(
-		&f.JUnitOutFile, "junitout", "junit.xml",
+		&f.JUnitOutFile, "junitout", "",
 		"Write the test output to a new file at this path as a JUnit file if running tests.")
 	flag.StringVar(
 		&f.RawTestOutFile, "rawtestout", "",
@@ -53,19 +53,26 @@ func (f *TestJSONFlags) RunTestCmd(cmdline []string) (err error) {
 	if f != nil {
 		if f.JUnitOutFile != "" {
 			var jf *os.File
+			if err := os.MkdirAll(filepath.Dir(f.JUnitOutFile), 0o755); err != nil {
+				return fmt.Errorf("failed to create directory for JUnit output: %v", err)
+			}
 			if jf, err = os.Create(f.JUnitOutFile); err != nil {
 				return err
 			}
 			defer func() {
 				err = errors.Join(err, jf.Close())
 			}()
-			writers = append(writers, json2junit.NewConverter(jf))
+			c := json2junit.NewConverter(jf)
+			defer func() {
+				err = errors.Join(err, c.Close())
+			}()
+			writers = append(writers, c)
 			needJSON = true
 		}
 		if f.RawTestOutFile != "" {
 			var rf *os.File
 			if err := os.MkdirAll(filepath.Dir(f.RawTestOutFile), 0o755); err != nil {
-				return fmt.Errorf("failed to create directory for raw test output: %w", err)
+				return fmt.Errorf("failed to create directory for raw test output: %v", err)
 			}
 			if rf, err = os.Create(f.RawTestOutFile); err != nil {
 				return err

--- a/eng/pipeline/stages/run-stage.yml
+++ b/eng/pipeline/stages/run-stage.yml
@@ -270,8 +270,8 @@ stages:
                       -builder '${{ parameters.builder.os }}-${{ parameters.builder.arch }}-${{ parameters.builder.config }}' `
                       $(if ('${{ parameters.builder.experiment }}') { '-experiment'; '${{ parameters.builder.experiment }}' }) `
                       $(if ('${{ parameters.builder.fips }}') { '-fipsmode' }) `
-                      -junitout '$(Build.SourcesDirectory)/eng/artifacts/TestResults.xml' `
-                      -rawtestout '$(Build.SourcesDirectory)/eng/artifacts/RawTestOutput/RawTestOutput-attempt-${{ attempt }}.txt'
+                      -junitout '$(Build.SourcesDirectory)/eng/artifacts/RawTestOutput/TestResults.xml' `
+                      -rawtestout '$(Build.SourcesDirectory)/eng/artifacts/RawTestOutput/raw-json-attempt-${{ attempt }}.txt'
                 ${{ if eq(length(parameters.retryAttempts), 1) }}:
                   displayName: Test
                 ${{ else }}:
@@ -286,17 +286,19 @@ stages:
                   # Don't fail the job if we are at the last retry and the builder is marked as broken.
                   continueOnError: true
 
+            # Publish test results. Use the last-overwritten result, if multiple attempts were made.
             - task: PublishTestResults@2
               displayName: Publish test results
               condition: succeededOrFailed()
               inputs:
                 testResultsFormat: JUnit
-                testResultsFiles: $(Build.SourcesDirectory)/eng/artifacts/TestResults.xml
+                testResultsFiles: $(Build.SourcesDirectory)/eng/artifacts/RawTestOutput/TestResults.xml
                 testRunTitle: $(System.JobDisplayName)
                 buildPlatform: ${{ parameters.builder.arch }}
                 buildConfiguration: ${{ parameters.builder.config }}
                 publishRunAttachments: true
 
+            # Publish raw test JSON output and the published (or attempted to publish) JUnit XML.
             - task: ${{ iif(eq(variables['System.TeamProject'], 'public'), '', '1ES.') }}PublishPipelineArtifact@1
               displayName: Publish raw test results
               condition: succeededOrFailed()

